### ISSUE TITLE
VS2015 Debugger wigs out - Locals window especially

### ISSUE THREAD

**Issue by Vitoc456** [Fody/PropertyChanged#167]
We've got a fairly large project that has been using PropertyChanged.Fody for the past several months without any detected issues.  We (I) recently ran across some very strange behavior in debugging in VS2015 and have tracked it down to PropertyChanged.Fody.  When it enters this bad state, the debugger Locals window starts showing all kinds of phantom entries as well as duplicate entries, some with different values.  Usually, the locals window only shows stuff as it comes into scope, but in this state it seems to almost preload all the variables used in the method.  You can't mouse over certain properties in the code window.  It also appears under certain circumstances conditional breakpoints don't work properly.  You always hit the breakpoint even when the specified criteria isn't met.  This coupled with the inability to easily use Linq extension methods ([noted here](https://github.com/Fody/Fody/issues/130)) is making for a borderline unusable debugging experience.

I'm attaching a zipped up solution exhibiting some of this behavior.  With a breakpoint on the very first line of the Class1.Blah method you'll see this behavior in the locals window.  It gets worse with more variables and more lines of code in the method.
[TestDebuggerUpdate2.zip](https://github.com/Fody/PropertyChanged/files/240929/TestDebuggerUpdate2.zip)

**Comment by Vitoc456** [comment]
![fodydebuggerstrange](https://cloud.githubusercontent.com/assets/12979856/14919477/45f9620a-0de6-11e6-9f43-379074541f0c.PNG)
For every conditional (if) it adds a blank variable in the locals window, and again, typically these only add to the locals window as they are declared and come into scope.

**Comment by SimonCropp** [comment]
can u please raise an issue on cecil https://github.com/jbevain/cecil

and include before and after pdbs and dlls from a simplified repro

**Comment by SimonCropp** [comment]
and link back to this issue

**Comment by Vitoc456** [comment]
Sorry, maybe I'm missing something.  Is Cecil related to Fody or does Fody have some sort of internal reference to it?  We aren't using Cecil.  I don't see any references to it anywhere, anyway.  Please see the zipped solution I uploaded to confirm.  I don't want to junk up Cecil github issues with stuff that's unrelated.  Cheers.

**Comment by SimonCropp** [comment]
@jbevain what do u think? does this sound like a cecil bug?

**Comment by jbevain** [comment]
@SimonCropp yes, side effect of two things:

* Those empty variables are «Debugger Hidden» variables, we're not serializing back those flags.
* We're not doing a great job at preserving the variable scopes, to say the least.

I've been working on it as part of the portable pdb support!

**Comment by Vitoc456** [comment]
Thanks for the responses so far.  I'm new to PropertyChanged, Fody, and "weaving" in general, so please bear with me.

@jbevain In your opinion, is it possible this might be related to why we're seeing multiple instances of the same variable in the locals window?  Also lots of NullReferences for variables in that window that aren't actually null?  Perhaps being unable to mouse over properties in the code window as well?  I'm having a hell of a time debugging this method.  It's as if the debugger is completely out of sync with the actual values in memory.  Objects aren't lining up.  The longer and more complex the method, the worse it gets.  Providing simple examples doesn't do it justice.

Here's an example of what I'm dealing with (heavily edited):
![propertychangeddebuggernightmare](https://cloud.githubusercontent.com/assets/12979856/15055073/dd710a26-12c7-11e6-9f0c-8d09e03cae0e.png)

I know the object on the line being debugged is not null.  However it doesn't show me anything if I mouse over the BalanceAmount in the code window.  It shows the object multiple times in the locals window as the green crossed out variables with a red X next to them and the "threw an exception of type 'System.NullReferenceException'" in the value column.  If you mouse over the value column in the Locals window it says "The value for this item is stale due to a problem that occurred while evaluating it. However your cursor over the refresh button for details. Old Value: 'blah' threw an exception of type 'System.NullReferenceException'".  I can't query the object's "BalanceAmount" in the Immediate window.  It complains about a different type completely not having a definition for BalanceAmount.  I have no visible way of getting to the actual value of the property through my debugging experience.  Quick watches don't work either.

**Comment by adminnz** [comment]
Is there any update to this? As I am also having this bug.

**Comment by Vitoc456** [comment]
From my end, no.  I got super lucky (I guess?) that I was able to resolve my business problem by optimizing the lengthy bugged method.  The debugger was next to worthless during the entire time -- most painful two weeks of debugging in my 15+ year career.  No joke.  Good luck!

**Comment by adminnz** [comment]
Well this bug makes the library pretty much useless because while it might save time writing code, it is completely wasted due to the extra time spent debugging.

**Comment by SimonCropp** [comment]
does it happen in both debug and release mode?

**Comment by adminnz** [comment]
Yes it happens in both debug and release mode (Release mode had Optimize Code unticked).

**Comment by SimonCropp** [comment]
@adminnz unfortunately this is a problem that needs to be fixed in cecil. and that is not something i currently have capacity to  work on. i suggest you ask @jbevain to spend some time on it or ask him how you can contribute to fixing it yourself.

the only option in the short term is to stop using this tool. 

i am sorry for any inconvenience this has caused

**Comment by ordinaryorange** [comment]
@SimonCropp got the same issue here. I've read the cecil issues, but cant decipher what the issue is or weather there is a temporary workaround or anything. If I'm correct it also looks like this has been an issue for a lot longer than this thread (circa 2010 ?), and @jbevain does not look like he is getting any time on a fix. Such a shame to have to bin the propertychanged tool for loss of debug functionality. Are there any workarounds at all ?

**Comment by jbevain** [comment]
This is something that we're working on. I just landed the support for the new portable pdb file format in Cecil's master, and we're finalizing a release. At that point in time, the workaround for users of VS 2015 Update 3 will simply be to change their csproj to tell the C# compiler to generate portable pdbs.

The portable pdb work has been the occasion to largely refactor Cecil's symbol story. Better support for old native pdbs will be 0.11.

**Comment by vincenormand** [comment]
I have tried changing the DebugType from "full" to "portable", but the solution does not compile anymore : "Fody: An unhandled exception occured: ..."
Would you have any other solution to resolve the problem, or an estimate for when a fix could be available?
Thank you. Here is the section i changed in the .csproj :

`<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|x86' ">
    <PlatformTarget>x86</PlatformTarget>
    <DebugSymbols>true</DebugSymbols>
    <DebugType>portable</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\Debug\</OutputPath>
    <DefineConstants>TRACE;DEBUG</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
    <GenerateSerializationAssemblies>Off</GenerateSerializationAssemblies>
  </PropertyGroup>`

**Comment by lademone** [comment]
@vincenormand Same issue here

**Comment by lambdanator** [comment]
@lademone @vincenormand  Same issue too..

**Comment by SimonCropp** [comment]
re the portable pdb. i am waiting on the next version of cecil that supports that format.

**Comment by vincenormand** [comment]
Temporarily, i compiled this branch and replaced the Mono.Cecil DLL files in the nuget package.
https://github.com/jbevain/cecil/tree/5a7322d0db79049faa0336df003c5ce0b111b395

It works now, but i will be waiting for the next release of Cecil, and then, a new PropertyChanged version.
Thank you

**Comment by ordinaryorange** [comment]
@vincenormand nice work, your fork is a sweet workaround. Very much appreciated.

**Comment by allanrsmith** [comment]
What fork?  I don't see a fork by @vincenormand.

**Comment by ordinaryorange** [comment]
@allanrsmith  loose use of the word fork. Download the branch from @vincenormand's link, compile it and replace the .dlls as mentioned with the newly compiled ones.

**Comment by vincenormand** [comment]
@SimonCropp 
Looks like something is finally comming!
https://github.com/jbevain/cecil/pull/351

**Comment by user1568891** [comment]
Any idea when the cecil version with the fix will be applied to fody?

**Comment by SimonCropp** [comment]
applied. can u try the newest nugets

**Comment by vincenormand** [comment]
I tested the new Nugets in Visual Studio 2015, but i think it is breaking the PDBs.
The debbugger is not working at all...

**Comment by SimonCropp** [comment]
@jbevain anything u can add here? any info we can give u?

**Comment by vincenormand** [comment]
After some researches, the problem was due to a Debbuging option in VS : "Use Managed Compatibility Mode".
With this option checked, the debbuger is not working in many classes, not all, but many. Especially in functions or methods called after a constructor.
After I unchecked it, it worked fine.

I don't think i need this option checked, so i will leave it unchecked. But it would be interesting to know why this option causes the bug.

**Comment by jbevain** [comment]
@SimonCropp that doesn't ring a bell no.
@vincenormand thanks for the investigation, I'll get in touch with the VS debugger folks and see what that option does.

**Comment by SimonCropp** [comment]
does this still happen in  2.0.0-beta001  ?

**Comment by SimonCropp** [comment]
and please use with Fody 2.0.10

**Comment by vincenormand** [comment]
This bug seems to be fixed with Fody version 2.0.10.
Thank you

**Comment by vincenormand** [comment]
Maybe this could be another thread, but i am asking it here anyway.

As of version 2.0.1, I see that the minimum .NET Framework version of PropertyChanged.Fody has been changed to 4.6.2.
Is there any particular reason for this?

I am asking because most of our applications that use PropertyChanged.Fody are using the .NET Framework 4.0. And because of this upgrade, we cannot update to the latest version of PropertyChanged.Fody.
And we are not going to install the framework 4.6.2 to all of our customers...

**Comment by cremor** [comment]
PropertyChanged.Fody v2.1.0, which is the latest version currently, supports .NET Standard 1.0, which is supported by version 4.5 of the .NET framework.

If you can't even target 4.5 you'll have to stay at PropertyChanged.Fody v1.53.0 which is the only version that supports Fody 2.0 but was before the switch to .NET standard.

**Comment by SimonCropp** [comment]
@vincenormand @cremor dicussions on supported runtimes should be handled in a new issue. given this seems to be fixed i will close
